### PR TITLE
ENH: Support scalar-first order of quaternion components in Rotation

### LINKED ISSUE #19649: ENH: as_quat() and from_quat() seams to be reverse x,y,z,w vs w,x,y,z
### Is your feature request related to a problem? Please describe.

I was trying to use the quaternion in scipy, and I found out that the format of the quaternion in scipy seams to be in a different order than the wiki page.

On your page:
https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.transform.Rotation.from_quat.html
You say:
"Each row is a (possibly non-unit norm) quaternion representing an active rotation, in scalar-last (x, y, z, w) format. Each quaternion will be normalized to unit norm."
but the example seams to have the scalar in front?
```
r = R.from_quat([1, 0, 0, 0])
r.as_quat()
array([1., 0., 0., 0.])
```

At the wiki link that you provide on your doc web page, https://en.wikipedia.org/wiki/Quaternions_and_spatial_rotation
, it also says that the format is w, x, y, z
Wikipedia page says:
"We can avoid this by using four Euclidean coordinates w, x, y, z, ..."
 
I am using a ABB robot and the documentation also says that the quaternion is in the format w, x, y, z.
see link: https://developercenter.robotstudio.com/api/robotstudio/api/ABB.Robotics.Math.Quaternion.html
q1, q2, q3, q4
q1 Scalar part: q1, Vector part: q2, q3, q4

### Describe the solution you'd like.

To add a configuration for the quaternion.
same as Euler angle that have the possibility to be configure as 'XYZ' or 'ZYX'
```
Rotation.from_quat(type cls, quat, conf = 'wxyz')
```
Without conf, from_quat could have by default 'xyzw', and this could be backward compatible.

### Describe alternatives you've considered.

_No response_

### Additional context (e.g. screenshots, GIFs)

English is not my native language, I could have made a few typo here and there.
Thanks and sorry

### PR DESCRIPTION
**Pull request by nmayorov**
#### Reference issue
Closes #19649

#### What does this implement/fix?
Adds optional arguments `scalar_first=False`  to `Rotation.from_quat` and `Rotation.as_quat` to support the alternative scalar-first quaternion order.

#### Additional information
The transformation between orders was implemented using `np.roll`, I haven't tried to implement it in low-level Cython. My reasoning was that there is already Python usage in these parts (highlighted by yellow in HTML by Cython).

I suspect that Cython usage might've become somewhat unreasonable in this file, rather inconvenient and hard to read or modify.

---

@scottshambaugh, would appreciate if you can review it.

### CODE REVIEW

File: scipy/spatial/transform/_rotation.pyx [hunk 1]
```diff
@@ -910,8 +915,12 @@ cdef class Rotation:
         ----------
         quat : array_like, shape (N, 4) or (4,)
             Each row is a (possibly non-unit norm) quaternion representing an
-            active rotation, in scalar-last (x, y, z, w) format. Each
-            quaternion will be normalized to unit norm.
+            active rotation. The assumed component order depends on `scalar_first`
```
Comment (scottshambaugh) [inline scipy/spatial/transform/_rotation.pyx:918]: I think this comment (and elsewhere) should be more explicit about which of the orders corresponds to `scalar_first = False`

File: scipy/spatial/transform/tests/test_rotation.py [hunk 2]
```diff
@@ -39,6 +39,52 @@ def test_from_single_2d_quaternion():
...
+    rng = np.random.RandomState(0)
+
+    r = Rotation.from_quat([1, 0, 0, 0], scalar_first=True)
+    assert_allclose(r.as_matrix(), np.eye(3), rtol=1e-15, atol=1e-16)
+
+    r = Rotation.from_quat(np.tile([1, 0, 0, 0], (10, 1)), scalar_first=True)
+    assert_allclose(r.as_matrix(), np.tile(np.eye(3), (10, 1, 1)),
+                    rtol=1e-15, atol=1e-16)
+
+    q = rng.randn(100, 4)
```
Comment (scottshambaugh) [inline scipy/spatial/transform/tests/test_rotation.py:52]: I'm not really following why all these extra tests are needed? Would think that you could just compare e.g.
`
assert(Rotation.from_quat([1, 2, 3, 4], scalar_first=True).as_quat(scalar_first=False), Rotation.from_quat([2, 3, 4, 1], scalar_first=False).as_quat(scalar_first=False))`

**Comment by scottshambaugh** [comment]
Left some comments, but to take a step back I think that we should mirror the API that exists for `from_euler`, etc where the keyword is for example `seq = XYZ`, instead of making a new keyword 'scalar_first'. In this case, we would have the keyword be `seq = 'xyzw' (default) or 'wxyz'`. While this is perhaps overkill versus a boolean for only having two options, this keeps API consistency and also is more explicit.

File: scipy/spatial/transform/tests/test_rotation.py (hunk 2 above)
Comment (nmayorov) [reply scipy/spatial/transform/tests/test_rotation.py:52]: Perhaps I've lost some skills in writing tests.

My reasoning was to keep tests as "unit" as possible and not couple new features in `as_quat` and `from_quat`.

Maybe it's an overkill here, but intuitively it makes sense to me. For example, what if I not implement anything for `scalar_first`? Then `as_quat` and `from_quat` will work consistently, but incorrectly.

**Comment by nmayorov** [comment]
> Left some comments, but to take a step back I think that we should mirror the API that exists for `from_euler`, etc where the keyword is for example `seq = 'xyz'`, instead of making a new keyword `scalar_first`. In this case, we would have the keyword be `seq = 'xyzw' (default) or 'wxyz'`. While this is perhaps overkill versus a boolean for only having two options, this keeps API consistency and also is more explicit.

I'm not entirely opposed to it, but not a fan of it. Sort of "+0", I'd like to hear other opinions.

I don't think the situation is very similar to Euler angles. There are only 2 conventions: scalar last and scalar first. And I haven't heard of "sequence of quaternion components", sounds a bit artificial.

From a programming experience, it's better to keep things simple and use bool for boolean values (more so for compiled languages, but still).

File: scipy/spatial/transform/_rotation.pyx (hunk 1 above)
Comment (nmayorov) [reply scipy/spatial/transform/_rotation.pyx:918]: For it we have to give some meaningful definition for the components or give an example for an identity rotation. 

I think the latter approach will be very appropriate and easy to understand. I'll work on that.

**Comment by lucascolley** [comment]
> > Left some comments, but to take a step back I think that we should mirror the API that exists for `from_euler`, etc where the keyword is for example `seq = 'xyz'`, instead of making a new keyword `scalar_first`. In this case, we would have the keyword be `seq = 'xyzw' (default) or 'wxyz'`. While this is perhaps overkill versus a boolean for only having two options, this keeps API consistency and also is more explicit.
> 
> I'm not entirely opposed to it, but not a fan of it. Sort of "+0", I'd like to hear other opinions.

@rtnhn @jmw182 @alspitz @patgadget any preference?

**Comment by nmayorov** [comment]
> Left some comments, but to take a step back I think that we should mirror the API that exists for `from_euler`, etc where the keyword is for example `seq = 'xyz'`, instead of making a new keyword `scalar_first`. In this case, we would have the keyword be `seq = 'xyzw' (default) or 'wxyz'`. While this is perhaps overkill versus a boolean for only having two options, this keeps API consistency and also is more explicit.

I'd be happier to use `order='xyzw'` - it makes more sense to me, "seq" is not a right choice in my opinion.

File: scipy/spatial/transform/_rotation.pyx (hunk 1 above)
Comment (nmayorov) [reply scipy/spatial/transform/_rotation.pyx]: I ended up writing definitions from angle and axis of a rotation. See if you like it.

I have to say, that the whole docstring of `Rotation` needs a revision. It's not very informative or cleanly written.

**Comment by jmw182** [comment]
> > > Left some comments, but to take a step back I think that we should mirror the API that exists for `from_euler`, etc where the keyword is for example `seq = 'xyz'`, instead of making a new keyword `scalar_first`. In this case, we would have the keyword be `seq = 'xyzw' (default) or 'wxyz'`. While this is perhaps overkill versus a boolean for only having two options, this keeps API consistency and also is more explicit.
> > 
> > 
> > I'm not entirely opposed to it, but not a fan of it. Sort of "+0", I'd like to hear other opinions.
> 
> @RTnhN @jmw182 @alspitz @patgadget any preference?

I don't have a strong preference, but a boolean input makes slightly more sense to me. To me, `seq=xyzw` or `order=xyzw` implies that any arbitrary order of those components is valid, e.g. `seq=zywx`. With a boolean it's clear that there are only two options.

**Comment by RTnhN** [comment]
> > > Left some comments, but to take a step back I think that we should mirror the API that exists for `from_euler`, etc where the keyword is for example `seq = 'xyz'`, instead of making a new keyword `scalar_first`. In this case, we would have the keyword be `seq = 'xyzw' (default) or 'wxyz'`. While this is perhaps overkill versus a boolean for only having two options, this keeps API consistency and also is more explicit.
> > 
> > 
> > I'm not entirely opposed to it, but not a fan of it. Sort of "+0", I'd like to hear other opinions.
> 
> @RTnhN @jmw182 @alspitz @patgadget any preference?

I am generally okay with either way for this. I do really like that the `order = 'wxyz'` is really explicit.

File: scipy/spatial/transform/tests/test_rotation.py (hunk 2 above)
Comment (nmayorov) [reply scipy/spatial/transform/tests/test_rotation.py:52]: I guess I've mostly missed your point.

Let me answer properly. The first pair of my tests are checking how identity quaternion looks in scalar-first format. 

The second pair works on a random input. And in this case using `np.roll` is inevitable.

I think my tests are fine, but I might have just used 2 simpler tests as your suggested (for a single and multiple quaternions).

**Comment by nmayorov** [comment]
So far looks like there is no strong preference. I also don't have a strong preference and in this case I'd keep it as it is now.

I think we can merge it rather quickly. Who already participates, please leave comments/reviews if you are interested.

@scottshambaugh please check my docstring updates.

**Comment by alspitz** [comment]
> > > Left some comments, but to take a step back I think that we should mirror the API that exists for `from_euler`, etc where the keyword is for example `seq = 'xyz'`, instead of making a new keyword `scalar_first`. In this case, we would have the keyword be `seq = 'xyzw' (default) or 'wxyz'`. While this is perhaps overkill versus a boolean for only having two options, this keeps API consistency and also is more explicit.
> > 
> > 
> > I'm not entirely opposed to it, but not a fan of it. Sort of "+0", I'd like to hear other opinions.
> 
> @RTnhN @jmw182 @alspitz @patgadget any preference?

I vote for boolean. It's not quite the same as Euler ordering (as mentioned arbitrary order doesn't make sense).

As far as the other conventions mentioned, handedness, active vs passive, etc., they seem to go beyond quaternions and generally affect rotation conventions so should probably leave them out of this API. E.g. the cross product in numpy is always right-handed.

**Comment by nmayorov** [comment]
If anyone can make some sort of conceptual approval, I'm ready to merge it.

I don't see what else should be changed for this feature to go into the main.

**Review by lucascolley — APPROVED** [review]
LGTM and seems to align with all of the discussion so far, thanks @nmayorov. Maybe give a few days in case @scottshambaugh has any further comments.

File: scipy/spatial/transform/_rotation.pyx [hunk 3]
```diff
@@ -1607,14 +1639,33 @@ cdef class Rotation:
...
+        The 4 components of a quaternion are divided into a scalar part ``w``
+        and a vector part ``(x, y, z)`` and can be expressed from the angle
+        ``theta`` and the axis ``n`` of a rotation as follows::
+
+            w = cos(theta / 2)
+            x = sin(theta / 2) * n_x
+            y = sin(theta / 2) * n_y
+            z = sin(theta / 2) * n_z
+
+        There are 2 conventions to oder the components in a quaternion:
```
Comment (scottshambaugh) [inline scipy/spatial/transform/_rotation.pyx:1657]:
```suggestion
        There are 2 conventions to order the components in a quaternion:
```

File: scipy/spatial/transform/_rotation.pyx [hunk 4]
```diff
@@ -1639,23 +1693,18 @@ cdef class Rotation:
         >>> from scipy.spatial.transform import Rotation as R
         >>> import numpy as np
 
-        Represent a single rotation:
+        A rotation can be represented as a quaternion with either scalar-last
+        (default) or scalar-first component order.
+        This is shown for a single rotaiton:
```
Comment (scottshambaugh) [inline scipy/spatial/transform/_rotation.pyx:1698]:
```suggestion
        This is shown for a single rotation:
```

File: scipy/spatial/transform/tests/test_rotation.py (hunk 2 above)
Comment (scottshambaugh) [reply scipy/spatial/transform/tests/test_rotation.py:52]: Got it, yeah I think the tests are fine as is

**Comment by scottshambaugh** [comment]
Two small typo corrections, but the docstrings are much improved. I agree they probably need a total overhaul - for example the vector definition here could be linked to the `as_rotvec` definition.

One more small nitpick is that it seems a bit roundabout to me to have the keyword default to `scalar_first = False` rather than to rename and set `scalar_last = True`. But don't feel too strongly about that, it's good as-is!

**Review by j-bowhay — COMMENTED** [review]
Non-blocking but should we consider making `scalar_first` keyword only as per https://docs.scipy.org/doc/scipy/dev/missing-bits.html#required-keyword-names

**Comment by nmayorov** [comment]
> One more small nitpick is that it seems a bit roundabout to me to have the keyword default to `scalar_first = False` rather than renamed to `scalar_last = True`. But don't feel too strongly about that, it's good as-is!

Should I change it? There is some reason to have a default value set to False (like "inactive") also it is similar to "canonical" argument. I think no one would notice that.

Typos are fixed.

> Non-blocking but should we consider making `scalar_first` keyword only as per https://docs.scipy.org/doc/scipy/dev/missing-bits.html#required-keyword-names

Sure, if it is considered a better approach now. Changed that.

**Review by lucascolley — APPROVED** [review]
Good catch @j-bowhay! I'm happy with this now, thanks for the review @scottshambaugh and the input all, and thanks @nmayorov !

**Comment by lucascolley** [comment]
bonus points if anyone wants to add a release note to the wiki ;)

**Comment by nmayorov** [comment]
I've added the note. Thanks everyone for the assistance.

**Pull request merged**